File: app/src/main/java/ilapin/opengl_research/domain/PlayerController.kt
package ilapin.opengl_research.domain

import kotlin.math.abs

/**
 * @author raynor on 07.02.20.
 */
class PlayerController(
    private val leftJoystick: Joystick,
    private val rightJoystick: Joystick
) {
    private var _movingFraction = 0f
    private var _strafingFraction = 0f
    private var _horizontalSteeringFraction = 0f
    private var _verticalSteeringFraction = 0f

    fun update() {
        _movingFraction = -(leftJoystick.position.y.takeIf { abs(it) >= THRESHOLD } ?: 0f)
        _strafingFraction = leftJoystick.position.x.takeIf { abs(it) >= THRESHOLD } ?: 0f

        _horizontalSteeringFraction = -(rightJoystick.position.x.takeIf { abs(it) >= THRESHOLD } ?: 0f)
        _verticalSteeringFraction = -(rightJoystick.position.y.takeIf { abs(it) >= THRESHOLD } ?: 0f)
    }

    val movingFraction: Float
        get() = _movingFraction

    val strafingFraction: Float
        get() = _strafingFraction

    val horizontalSteeringFraction: Float
        get() = _horizontalSteeringFraction

    val verticalSteeringFraction: Float
        get() = _verticalSteeringFraction

    companion object {

        private const val THRESHOLD = 0.01f
    }
}

File: app/src/main/assets/scenes/player_controller.js
function PlayerController(leftJoystickController, rightJoystickController) {

    this.THRESHOLD = 0.01

    this.leftJoystickController = leftJoystickController;
    this.rightJoystickController = rightJoystickController;

    this.movingFraction = 0;
    this.strafingFraction = 0;
    this.horizontalSteeringFraction = 0;
    this.verticalSteeringFraction = 0;

    this.update = function() {
        if (Math.abs(this.leftJoystickController.joystickPositionY) >= this.THRESHOLD) {
            this.movingFraction = this.leftJoystickController.joystickPositionY;
        } else {
            this.movingFraction = 0;
        }
        if (Math.abs(this.leftJoystickController.joystickPositionX) >= this.THRESHOLD) {
            this.strafingFraction = this.leftJoystickController.joystickPositionX;
        } else {
            this.strafingFraction = 0;
        }

        if (Math.abs(this.rightJoystickController.joystickPositionX) >= this.THRESHOLD) {
            this.horizontalSteeringFraction = -this.rightJoystickController.joystickPositionX;
        } else {
            this.horizontalSteeringFraction = 0;
        }
        if (Math.abs(this.rightJoystickController.joystickPositionY) >= this.THRESHOLD) {
            this.verticalSteeringFraction = -this.rightJoystickController.joystickPositionY;
        } else {
            this.verticalSteeringFraction = 0;
        }
    }
}